### DISCUSSION TITLE
Get select with options (selectinload) using response schema

### DISCUSSION THREAD

**Discussion by ProgrammingStore** [fastapi/sqlmodel#1527] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel

class Parent(SQLModel, table=True):
    id: UUID = sm.Field(UUID, primary_key=True)
    childs:List[Child]= sm.Relationship(
        back_populates="parent"
    )

class Child(SQLModel, table=True):
    parent_id:UUID=sm.Field()
        sa_column=sm.Column(
            sm.ForeignKey("parentr.id")
    )
    parent: "Parent" = sm.Relationship(
        back_populates="childs"
    )

#read schemas

class IChildRead(BaseModel):
    id:UUID

class IParentReadWithChilds(BaseModel):
    childs:List[IChildRead]
```

### Description

What i want to get from pydantic response schema ? I want to get a query with select lazy options, using the schema of response, because this information already contains in relationships of pydantic models.
For example:
select(Parent).options(selectinload(Parent.Childs)),

The existance of property in pydantic model gives information about the need to use selectinload. Is there any solutions for it? 

### Operating System

Linux

### Operating System Details

any

### SQLModel Version

any

### Python Version

any

### Additional Context

any

**Comment by cycledriver** [comment]
There are 2 ways to Child load with a selectin (or whatever type of lazy option):
1) Define the lazyload option in your model:

``` python
    childs:List[Child]= sm.Relationship(
        back_populates="parent",
        sa_relationship_kwargs: {"lazyload": "selectin"},
    )
```

Whenever you select Parent objects, the children will **always** be loaded.

2) At query time

If you want to be a bit more selective about when you do the selectin load, you can add it to the query:

``` python
from sqlalchemy.orm import selectinload

all_parents = session.exec(select(Parent).options(selectinload(Parent.childs)).all()
```

Pretty much anything you can do with the loads is described here:

https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html

**Comment by Bewinxed** [comment]
> There are 2 ways to Child load with a selectin (or whatever type of lazy option):
> 
> 1. Define the lazyload option in your model:
> 
> ```python
>     childs:List[Child]= sm.Relationship(
>         back_populates="parent",
>         sa_relationship_kwargs: {"lazyload": "selectin"},
>     )
> ```
> 
> Whenever you select Parent objects, the children will **always** be loaded.
> 
> 2. At query time
> 
> If you want to be a bit more selective about when you do the selectin load, you can add it to the query:
> 
> ```python
> from sqlalchemy.orm import selectinload
> 
> all_parents = session.exec(select(Parent).options(selectinload(Parent.childs)).all()
> ```
> 
> Pretty much anything you can do with the loads is described here:
> 
> https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html

Any way to use this without type errors? I'm getting this complaint when I use the selectinload

```
No overloads for "exec" match the provided argumentsPylance[reportCallIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportCallIssue)
session.py(50, 15): Overload 2 is the closest match
Argument of type "List[PersonaTrait]" cannot be assigned to parameter "keys" of type "_AttrType" in function "selectinload"
  Type "List[PersonaTrait]" is incompatible with type "_AttrType"
    "List[PersonaTrait]" is incompatible with "QueryableAttribute[Any]"
    "List[PersonaTrait]" is incompatible with type "Literal['*']"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```

**Comment by Vr3n** [comment, marked as answer]
In version `0.0.21` Passing `lazy` instead of `lazyload` worked for me.

```python
 childs:List[Child]= sm.Relationship(
        back_populates="parent",
        sa_relationship_kwargs: {"lazy": "selectin"}, # lazy instead of lazyload
    )
```

**Comment by koldakov** [comment]
> There are 2 ways to Child load with a selectin (or whatever type of lazy option):
> 
> 1. Define the lazyload option in your model:
> 
> ```python
>     childs:List[Child]= sm.Relationship(
>         back_populates="parent",
>         sa_relationship_kwargs: {"lazyload": "selectin"},
>     )
> ```
> 
> Whenever you select Parent objects, the children will **always** be loaded.
> 
> 2. At query time
> 
> If you want to be a bit more selective about when you do the selectin load, you can add it to the query:
> 
> ```python
> from sqlalchemy.orm import selectinload
> 
> all_parents = session.exec(select(Parent).options(selectinload(Parent.childs)).all()
> ```
> 
> Pretty much anything you can do with the loads is described here:
> 
> https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html

@cycledriver Thank you for your reply! But I have related question, in that case all childs will be loaded, what can be a problem if there are a lot of childs. Do you know if there is a way to limit childs, or add where condition or whatever to load only N childs?

**Comment by koldakov** [comment]
> > There are 2 ways to Child load with a selectin (or whatever type of lazy option):
> > 
> > 1. Define the lazyload option in your model:
> > 
> > ```python
> >     childs:List[Child]= sm.Relationship(
> >         back_populates="parent",
> >         sa_relationship_kwargs: {"lazyload": "selectin"},
> >     )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Whenever you select Parent objects, the children will **always** be loaded.
> > 
> > 2. At query time
> > 
> > If you want to be a bit more selective about when you do the selectin load, you can add it to the query:
> > ```python
> > from sqlalchemy.orm import selectinload
> > 
> > all_parents = session.exec(select(Parent).options(selectinload(Parent.childs)).all()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Pretty much anything you can do with the loads is described here:
> > https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html
> 
> @cycledriver Thank you for your reply! But I have related question, in that case all childs will be loaded, what can be a problem if there are a lot of childs. Do you know if there is a way to limit childs, or add where condition or whatever to load only N childs?

Actually found a way:

`all_parents = session.exec(select(Parent).options(selectinload(Parent.childs), with_loader_criteria(Child, Forum.id < N)).all()`

**Comment by chriscarrollsmith** [comment]
> Any way to use this without type errors? I'm getting this complaint when I use the selectinload
> 
> ```
> No overloads for "exec" match the provided argumentsPylance[reportCallIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportCallIssue)
> session.py(50, 15): Overload 2 is the closest match
> Argument of type "List[PersonaTrait]" cannot be assigned to parameter "keys" of type "_AttrType" in function "selectinload"
>   Type "List[PersonaTrait]" is incompatible with type "_AttrType"
>     "List[PersonaTrait]" is incompatible with "QueryableAttribute[Any]"
>     "List[PersonaTrait]" is incompatible with type "Literal['*']"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
> ```

You can wrap your relationship's type declaration in `Mapped[]`:

```python
persona_traits: Mapped[List[PersonaTrait]] = Relationship(back_populates="persona")
```

That will shut Pyright up.

**Reply by PAVaillancourt-IvadoLabs** [reply]
For those searching: 
`from sqlalchemy.orm import Mapped`

**Reply by Raymondo97** [reply]
I brought this up in a similar discussion: #1853. 

The only issue with using Mapped is that it seems that SQLModel’s goal is to eliminate some of that boilerplate code that SQLAlchemy had. And SQLModel uses the Python types for interpreting the Pydantic and database column types. I’ve read that some people will add a TYPE_CHECKING flag (from Pydantic) to use it just for silencing the IDE, without affecting runtime.
